flang/lib/Parser/preprocessor.cpp

```diff
@@ -46,6 +49,110 @@ bool Definition::set_isDisabled(bool disable) {
   return was;
 }
 
+void Definition::Print(
+    llvm::raw_ostream &out, llvm::StringRef macroName) const {
+  if (isDisabled_) {
+    return;
+  }
+  if (!isFunctionLike_) {
+    // If it's not a function-like macro, then just print the replacement.
+    out << ' ' << replacement_.ToString();
+    return;
+  }
+
+  // The sequence of characters from which argument names will be created.
+  static llvm::StringRef charSeq{"ABCDEFGHIJKLMNOPQRSTUVWXYZ"};
+
+  auto couldCollide = [&](llvm::StringRef str) {
+    return !str.empty() && llvm::all_of(str, [&](char c) {
+      return charSeq.find(c) != llvm::StringRef::npos;
+    });
+  };
+
+  // For function-like macros we need to invent valid argument names (they
+  // are represented as ~A, ~B, ...). These invented names cannot collide
+  // with any other tokens in the macro definitions.
+  llvm::SmallSet<std::string, 10> usedNames;
+  for (size_t i{0}, e{replacement_.SizeInTokens()}; i != e; ++i) {
+    std::string tok{replacement_.TokenAt(i).ToString()};
+    if (tok.empty()) {
+      continue;
+    }
+    // The generated names will only use characters from `charSeq`, so
```

Comment:
You can avoid all of this complexity if you modify the preprocessor to retain the original actual argument names in the macro bodies.

Reply:
Yes, but I didn't want to make that change on my own.  Do you want me to go ahead with that?

Reply:
Do you want to store the argument list (with names) as well?  If not, we'd have to extract the names from the body, and sort them it based on the `~?~` character.

Reply:
I did something a bit different: I'm storing the argument names instead of making any changes to the tokenization.  The reason is that if there is a macro with an unused argument, we still need to give it some name.  Storing all of the actual names would completely eliminate the need to do that.